File: adapters/ts/oceanic_adapter.ts
/**
 * TypeScript Oceanic IR Adapter for executing and validating contracts.
 */

export interface OceanicIrPayload {
  kind: string;
  intent: string;
  evidence: string;
  proof_required?: boolean;
  target?: string;
  dissent_tokens?: string[];
  metadata?: Record<string, any>;
}

export interface AdapterResult {
  status: 'executed' | 'rejected' | 'error';
  adapter: 'typescript';
  intent?: string;
  evidence?: string;
  verifiedInAdapter?: boolean;
  timestamp?: string;
  reason?: string;
  message?: string;
}

export class TypeScriptOceanicAdapter {
  private context: Record<string, any>;

  constructor(context: Record<string, any> = {}) {
    this.context = context;
  }

  public executeIr(irPayload: OceanicIrPayload): AdapterResult {
    if (!irPayload || irPayload.kind !== 'oceanic_ir') {
      return { status: 'error', adapter: 'typescript', message: 'Invalid IR payload: wrong kind' };
    }

    const intent = irPayload.intent || '';
    const evidence = irPayload.evidence || '';

    if (!intent.trim()) {
      return { status: 'rejected', adapter: 'typescript', reason: 'Empty intent payload' };
    }

    return {
      status: 'executed',
      adapter: 'typescript',
      intent: intent,
      evidence: evidence,
      verifiedInAdapter: Boolean(evidence.trim()),
      timestamp: new Date().toISOString(),
    };
  }
}

File: adapters/rust/oceanic_adapter.rs
//! Rust Oceanic IR Adapter for executing and validating contracts.

use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct IrPayload {
    pub kind: String,
    pub intent: String,
    pub evidence: String,
    pub proof_required: bool,
    pub target: String,
    pub dissent_tokens: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct AdapterResult {
    pub status: String,
    pub adapter: String,
    pub intent: String,
    pub evidence: String,
    pub verified_in_adapter: bool,
    pub error: Option<String>,
}

pub struct RustOceanicAdapter {
    pub context: HashMap<String, String>,
}

impl RustOceanicAdapter {
    pub fn new(context: HashMap<String, String>) -> Self {
        Self { context }
    }

    pub fn execute_ir(&self, ir: &IrPayload) -> Result<AdapterResult, String> {
        if ir.kind != "oceanic_ir" {
            return Err("Invalid IR payload: wrong kind".to_string());
        }

        let intent = ir.intent.trim();
        let evidence = ir.evidence.trim();

        if intent.is_empty() {
            return Err("Empty intent payload".to_string());
        }

        Ok(AdapterResult {
            status: "executed".to_string(),
            adapter: "rust".to_string(),
            intent: intent.to_string(),
            evidence: evidence.to_string(),
            verified_in_adapter: !evidence.is_empty(),
            error: None,
        })
    }
}
